### PR TITLE
Fix CopyableText overflow

### PR DESCRIPTION
**Pull request by pwz266266**
Type: defect
Fixes: https://github.com/vector-im/element-web/issues/21040
Notes: Make the CopyableText scrollable so that longer text won't overflow, scroll bar is hidden.

## Before:
image

## After:
image

Signed-off-by: wenzhe <pwz26626664@gmail.com>

---
Here's what your changelog entry will look like:

## 🐛 Bug Fixes
 * Make the CopyableText scrollable so that longer text won't overflow, scroll bar is hidden. ([\#7851](https://github.com/matrix-org/matrix-react-sdk/pull/7851)). Fixes vector-im/element-web#21040. Contributed by @pwz266266.

Preview: https://pr7851--matrix-react-sdk.netlify.app
⚠️ Do you trust the author of this PR? Maybe this build will steal your keys or give you malware. Exercise caution. Use test accounts.

### CODE REVIEW

**Review by SimonBrandner — DISMISSED** [review]
LGTM codewise - thanks!

**Review by andybalaam — DISMISSED** [review]
Looks good, thank you!

**Review by t3chguy — CHANGES_REQUESTED** [review]
The scrollbar should not be missing, that makes the accessibility poor, the text should also have an ellipsis

File: res/css/views/elements/_CopyableText.scss [hunk 1]
```diff
@@ -26,7 +26,19 @@ limitations under the License.
...
+            display: none;
+        }
+    }
+
     .mx_CopyableText_copyButton {
+        position: relative;
+        right: 10px;
         flex-shrink: 0;
         width: 20px;
         height: 20px;
```
Comment (t3chguy) [inline res/css/views/elements/_CopyableText.scss:44]:
```suggestion
    .mx_CopyableText_innerText {
        overflow: auto;
    }

    .mx_CopyableText_copyButton {
        flex-shrink: 0;
```

And also get rid of the `cursor`, `margin-left` and `display` rules below too, they're superfluous please

**Comment by pwz266266** [comment]
Adapt changes from [t3chguy](https://github.com/t3chguy). LGTM

**Review by t3chguy — CHANGES_REQUESTED** [review]
Could we get a `margin-right` on `.mx_CopyableText_innerText` just to bring the button a bit further away. The css I told you to remove had a margin-left of 20px which only worked when the text wasn't overflowing, putting the margin on the innerText would work better, I think 16px is plenty though!

image

**Comment by pwz266266** [comment]
> Could we get a `margin-right` on `.mx_CopyableText_innerText` just to bring the button a bit further away. The css I told you to remove had a margin-left of 20px which only worked when the text wasn't overflowing, putting the margin on the innerText would work better, I think 16px is plenty though!
> 
> image

How about margin-left of 10px? This could bring the same amount of white space around the copy button since the outter CopyableText itself has a padding of 10px.

**Comment by t3chguy** [comment]
Hmm, playing with it some more I wonder whether it should just have `white-space: break-word;` instead, to make it span multiple lines

**Comment by pwz266266** [comment]
> Hmm, playing with it some more I wonder whether it should just have `white-space: break-word;` instead, to make it span multiple lines - what do you think?

I think both make sense, more of a design choice I guess?
Having multiple lines would result in:
image

I also checked other places using this component, should work fine too.
Code pushed which uses `word-break: break-all;`

**Comment by t3chguy** [comment]
@matrix-org/design the three options here are

Wrap:
image

Scroll:
image

Ellipsis:
image

**Review by janogarcia — COMMENTED** [review]
@t3chguy Given that it seems to be collapsed by default and thus not taking up any space in the screen until it's revealed, I'd recommend going for the **Wrap** variant, as long as it doesn't take way too many lines when revealed (let's say more than 20 lines).

---

As an alternate case:

Suppose we had to show similar inputs by default (uncollapsed) in other places where we need to save some space. In that case, I'd recommend using the ellipsis variant so that when clicking on the text, the ellipsis is replaced by a hard cut-off (no horizontal scroll), and the whole string is selected for you.

Frame 6502

**Comment by t3chguy** [comment]
@janogarcia this is a re-usable component

image

Used in many places already, so the decision here will impact all those. So would you say single-line ellipsis then? In future we could add an expand mechanism if we deem it useful.
Today it just sprawls out, so any decision is better than today's implementation

**Comment by janogarcia** [comment]
> So would you say single-line ellipsis then?

@t3chguy Makes sense. Would it be possible to implement the on click/selected behavior explained above (please refer to mockup)?

**Comment by t3chguy** [comment]
@janogarcia yup that should be reasonable

**Comment by t3chguy** [comment]
@pwz266266 are you up for making the changes outlined above?

**Comment by janogarcia** [comment]
@t3chguy Turns out that (in an unsurprising way) we have somewhat different renderings/treatments for that component.

**Settings**

We do have both multiline (`Versions`) and single line containers (`Access Token`).

**URLs**

- Default state: Ellipsized if overflowing.
- Selected state: The whole string is selected automatically on clicking on it, but the ellipsis char is not removed.

My recommendation would be implementing the [suggested behavior](https://github.com/matrix-org/matrix-react-sdk/pull/7851#pullrequestreview-969116145) if single line (auto-select + ellipsis that is removed on select), and just applying the auto-select behavior if multi line. It would require performing some updates to the component to have a more robust and consistent one.

---

settings

share

share-selected

**Comment by t3chguy** [comment]
@janogarcia great point, to clarify I was only specifically meaning we tweak the behaviour of CopyableText elements which contain text without any new lines (`\n`) characters.

**Comment by janogarcia** [comment]
@t3chguy Sure, but given that both cases are so tightly related, it would be a pity to normalize the behavior just for one case (no new lines) while keeping the other one working slightly inconsistently.

To recap, by normalizing the component we actually refer to two separate behaviors:

- How **text auto-selection** should work: On clicking the text, the whole string gets selected. Applies to both single line and multiple line variants. 
- How **overflowing text** should be handled: Show ellipsis if overflowing. Remove ellipsis and cut off text on clicking/selecting it. Only applies to single line variant.

I'd suggest that we normalize both behaviors, text auto-selection and overflow text.

If the former needs to happen in a separate PR because of scope creeping that's OK but, _if possible_, we should try addressing both at the same time as they are quite interdependent.

**Comment by janogarcia** [comment]
Sometimes we try to keep PRs super atomic, which isn't necessarily a bad thing, but that can also prevent us from introducing/performing more relevant, necessary updates right away once we realize the actual underlying issues.

It's an incremental improvement, I totally agree, but it can sometimes make dealing with UX debt more tedious (splitting the discussion, etc).

**Comment by t3chguy** [comment]
Perfect, thanks for the recap/summary

**Review by t3chguy — CHANGES_REQUESTED** [review]
See above, let me know if you need any help at all

**Pull request closed without merging**